### PR TITLE
[Enhancement] Restrict automatic partition expressions

### LINKED ISSUE #11444: Improve partition management
## Enhancement
Many databases support the function of importing and automatically creating partitions. We can support this function to complete the creation of partitions, and then use ttl to control the deletion of partitions. This ease of use will be better than the previous method.

### PR DESCRIPTION
**Pull request by Astralidea**
## What type of PR is this：
- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：

Fixes #11444

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto backported to target branch
  - [x] 3.0
  - [ ] 2.5
  - [ ] 2.4
  - [ ] 2.3

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 36 / 41 (87.80%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|36|41|87.80%|[645, 680, 688, 712, 713]|

File: fe/fe-core/src/main/java/com/starrocks/sql/parser/AstBuilder.java [hunk 1]
```diff
@@ -616,25 +623,45 @@ private List<String> checkAndExtractPartitionCol(FunctionCallExpr expr) {
                 throw new ParsingException(PARSER_ERROR_MSG.unsupportedExprWithInfo(expr.toSql(), "PARTITION BY"), pos);
             }
             Expr firstExpr = paramsExpr.get(0);
+            Expr secondExpr = paramsExpr.get(1);
+            String partitionName;
+            if (secondExpr instanceof SlotRef) {
+                partitionName = ((SlotRef) secondExpr).getColumnName();
```
Comment (ABingHuang) [inline fe/fe-core/src/main/java/com/starrocks/sql/parser/AstBuilder.java:629]:
```suggestion
                partitionColumnName = ((SlotRef) secondExpr).getColumnName();
```

File: fe/fe-core/src/main/java/com/starrocks/sql/parser/AstBuilder.java [hunk 2]
```diff
@@ -616,25 +623,45 @@ private List<String> checkAndExtractPartitionCol(FunctionCallExpr expr) {
...
-                columnList.add(((SlotRef) secondExpr).getColumnName());
-            } else {
+
+        } else if (FunctionSet.TIME_SLICE.equals(functionName)) {
+            if (paramsExpr.size() != 4) {
                 throw new ParsingException(PARSER_ERROR_MSG.unsupportedExprWithInfo(expr.toSql(), "PARTITION BY"), pos);
             }
-        } else if (FunctionSet.TIME_SLICE.equals(functionName)) {
             Expr firstExpr = paramsExpr.get(0);
+            String partitionName;
```
Comment (ABingHuang) [inline fe/fe-core/src/main/java/com/starrocks/sql/parser/AstBuilder.java:661]:
```suggestion
            String partitionColumnName;
```

File: fe/fe-core/src/main/java/com/starrocks/sql/parser/AstBuilder.java (hunk 1 above)
Comment (ABingHuang) [reply fe/fe-core/src/main/java/com/starrocks/sql/parser/AstBuilder.java:629]: better to put the check logic of DATA_TRUNC and TIME_SLICE into a function each.

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 34 / 38 (89.47%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|34|38|89.47%|[669, 694, 708, 709]|

**Review by ABingHuang — APPROVED** [review]
LGTM

**Review by gengjun-git — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by Astralidea** [comment]
run starrocks_admit_test

**Comment by Astralidea** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 34 / 38 (89.47%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|34|38|89.47%|[669, 694, 708, 709]|

**Comment by Astralidea** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-3.0

**Pull request merged**